// react-context-api/dist/main.cjs.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0

THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.

See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
}

var extendStatics$1=Object.setPrototypeOf||{__proto__:[]}instanceof Array&&function(e,t){e.__proto__=t;}||function(e,t){for(var r in t)t.hasOwnProperty(r)&&(e[r]=t[r]);};function __extends$1(e,t){function r(){this.constructor=e;}extendStatics$1(e,t),e.prototype=null===t?Object.create(t):(r.prototype=t.prototype,new r);}var __assign$1=Object.assign||function(e){for(var t,r=1,n=arguments.length;n>r;r++)for(var o in t=arguments[r])Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o]);return e};function __rest$1(e,t){var r={};for(var n in e)Object.prototype.hasOwnProperty.call(e,n)&&0>t.indexOf(n)&&(r[n]=e[n]);if(null!=e&&"function"==typeof Object.getOwnPropertySymbols){var o=0;for(n=Object.getOwnPropertySymbols(e);n.length>o;o++)0>t.indexOf(n[o])&&(r[n[o]]=e[n[o]]);}return r}function unwrapExports(e){return e&&e.__esModule&&Object.prototype.hasOwnProperty.call(e,"default")?e.default:e}function createCommonjsModule(e,t){return e(t={exports:{}},t.exports),t.exports}var hoistNonReactStatics=createCommonjsModule(function(e,t){var r,n,o,a,s,i,p,c;e.exports=(r={childContextTypes:!0,contextTypes:!0,defaultProps:!0,displayName:!0,getDefaultProps:!0,getDerivedStateFromProps:!0,mixins:!0,propTypes:!0,type:!0},n={name:!0,length:!0,prototype:!0,caller:!0,callee:!0,arguments:!0,arity:!0},o=Object.defineProperty,a=Object.getOwnPropertyNames,s=Object.getOwnPropertySymbols,i=Object.getOwnPropertyDescriptor,c=(p=Object.getPrototypeOf)&&p(Object),function e(t,u,l){if("string"!=typeof u){if(c){var f=p(u);f&&f!==c&&e(t,f,l);}var d=a(u);s&&(d=d.concat(s(u)));for(var y=0;d.length>y;++y){var m=d[y];if(!(r[m]||n[m]||l&&l[m])){var v=i(u,m);try{o(t,m,v);}catch(e){}}}return t}return t});}),getDisplayName_1=createCommonjsModule(function(e,t){Object.defineProperty(t,"__esModule",{value:!0}),t.default=function(e){return e.displayName||e.name||("string"==typeof e&&e.length>0?e:"Unknown")};}),getDisplayName=unwrapExports(getDisplayName_1),values=Object.values,keys=Object.keys,assign=Object.assign;function omit(e,t){return keys(t).filter(function(t){return -1===e.indexOf(t)}).reduce(function(e,r){return __assign$1({},e,((n={})[r]=t[r],n));var n;},{})}function prop(e,t){return t[e]}var isFn=function(e){return !!e&&"function"==typeof e},isValidRenderProp=function(e){return React__default.isValidElement(e)||isFn(e)};function adopt(e,t){if(!values(e).some(isValidRenderProp))throw Error("The render props object mapper just accept valid elements as value");var r=keys(e),n=function(e){var t=e.render,r=e.children,n=__rest$1(e,["render","children"]);return t&&isFn(t)?t(n):r&&isFn(r)&&r(n)};n.displayName="Adopt";return r.reduce(function(n,o,a){var s=prop(o,e),i=getDisplayName(n),p=getDisplayName(s),c=a===r.length-1,u=function(e){var r=e.render,a=e.children,i=__rest$1(e,["render","children"]);return React__default.createElement(n,__assign$1({},i),function(e){var n=omit(keys(i),e),p=r&&isFn(r)?r:a,u=function(e){var r,a=assign({},n,((r={})[o]=e,r)),s=t&&isFn(t)&&c?t(a):a;return p&&isFn(p)?p(s):null};return isFn(s)?React__default.createElement(s,assign({},i,e,{render:u})):React__default.cloneElement(s,{},u)})};return u.displayName=i+"("+p+")",isFn(s)?hoistNonReactStatics(u,s):u},n)}var Adopt=function(e){function t(t){var r=e.call(this,t)||this;return r.Composed=adopt(t.mapper,r.props.mapProps),r}return __extends$1(t,e),t.prototype.render=function(){var e=this.props,t=__rest$1(e,["mapper","mapProps"]);return React__default.createElement(this.Composed,__assign$1({},t))},t}(React__default.Component);

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

function createContextAPI(contextList) {
    /**
     * ContextProvider
     */
    var ContextProvider = function (_React$Component) {
        inherits(ContextProvider, _React$Component);

        function ContextProvider() {
            classCallCheck(this, ContextProvider);
            return possibleConstructorReturn(this, (ContextProvider.__proto__ || Object.getPrototypeOf(ContextProvider)).apply(this, arguments));
        }

        createClass(ContextProvider, [{
            key: "render",
            value: function render() {
                var _this2 = this;

                return Object.values(contextList).reduce(function (Parent, cx) {
                    return React.createElement(cx.Provider, null, React.cloneElement(Parent, Object.assign({}, _this2.props)));
                }, this.props.children);
            }
        }]);
        return ContextProvider;
    }(React.Component);

    var withContextProvider = function withContextProvider(Base) {
        var WithContextProvider = function (_React$Component2) {
            inherits(WithContextProvider, _React$Component2);

            function WithContextProvider() {
                classCallCheck(this, WithContextProvider);
                return possibleConstructorReturn(this, (WithContextProvider.__proto__ || Object.getPrototypeOf(WithContextProvider)).apply(this, arguments));
            }

            createClass(WithContextProvider, [{
                key: "render",
                value: function render() {
                    var _a = this.props,
                        forwardedRef = _a.forwardedRef,
                        rest = __rest(_a, ["forwardedRef"]);
                    return React.createElement(ContextProvider, null, React.createElement(Base, Object.assign({ ref: forwardedRef }, rest)));
                }
            }]);
            return WithContextProvider;
        }(React.Component);

        WithContextProvider.displayName = "WithContextProvider";
        return React.forwardRef(function (props, ref) {
            return React.createElement(WithContextProvider, Object.assign({}, props, { forwardedRef: ref }));
        });
    };
    /**
     * ContextConsumer
     */
    var getContextConsumer = function getContextConsumer(contextToRenderProps) {
        var pickedContextList = contextList;
        if (contextToRenderProps && contextToRenderProps.length > 0) {
            pickedContextList = contextToRenderProps.reduce(function (acc, cxKey) {
                if (contextList[cxKey]) acc[cxKey] = contextList[cxKey];
                return acc;
            }, {});
        }
        return adopt(Object.keys(pickedContextList).reduce(function (acc, cxKey) {
            var PatialConsumer = pickedContextList[cxKey].Consumer;
            acc[cxKey] = React.createElement(PatialConsumer, null);
            return acc;
        }, {}));
    };

    var ContextConsumer = function (_React$Component3) {
        inherits(ContextConsumer, _React$Component3);

        function ContextConsumer() {
            classCallCheck(this, ContextConsumer);
            return possibleConstructorReturn(this, (ContextConsumer.__proto__ || Object.getPrototypeOf(ContextConsumer)).apply(this, arguments));
        }

        createClass(ContextConsumer, [{
            key: "render",
            value: function render() {
                var Consumer = getContextConsumer(this.props.contextToRenderProps);
                return React.createElement(Consumer, null, this.props.children);
            }
        }]);
        return ContextConsumer;
    }(React.Component);

    var withContextConsumer = function withContextConsumer(contextToRenderProps) {
        return function (Base) {
            var Consumer = getContextConsumer(contextToRenderProps);

            var WithContextConsumer = function (_React$Component4) {
                inherits(WithContextConsumer, _React$Component4);

                function WithContextConsumer() {
                    classCallCheck(this, WithContextConsumer);
                    return possibleConstructorReturn(this, (WithContextConsumer.__proto__ || Object.getPrototypeOf(WithContextConsumer)).apply(this, arguments));
                }

                createClass(WithContextConsumer, [{
                    key: "render",
                    value: function render() {
                        var _a = this.props,
                            forwardedRef = _a.forwardedRef,
                            rest = __rest(_a, ["forwardedRef"]);
                        return React.createElement(Consumer, null, function (data) {
                            return React.createElement(Base, Object.assign({ ref: forwardedRef }, data, rest));
                        });
                    }
                }]);
                return WithContextConsumer;
            }(React.Component);

            WithContextConsumer.displayName = "WithContextConsumer";
            return React.forwardRef(function (props, ref) {
                return React.createElement(WithContextConsumer, Object.assign({}, props, { forwardedRef: ref }));
            });
        };
    };
    return {
        ContextProvider: ContextProvider,
        ContextConsumer: ContextConsumer,
        getContextConsumer: getContextConsumer,
        withContextProvider: withContextProvider,
        withContextConsumer: withContextConsumer
    };
}

module.exports = createContextAPI;
